src/Json_decode.ml

```diff
@@ -184,14 +184,17 @@ let optional decode json =
   try Some (decode json) with
   | DecodeError _ -> None
 
-let rec oneOf decoders json =
-  match decoders with
-  | [] ->
-    let length = List.length decoders in
-    raise @@ DecodeError ({j|Expected oneOf $length, got |j} ^ _stringify json)
-  | decode :: rest ->
-    try decode json with
-    | DecodeError _ -> oneOf rest json
+let oneOf decoders json =
+  let rec inner decoders errors =
+    match decoders with
+    | [] ->
+        raise @@ DecodeError
+              ({j|All decoders given to oneOf failed. Here are all the errors: $errors. And the JSON being decoded: |j} ^ _stringify json)
+    | decode::rest ->
+        try decode json with
+        | DecodeError e ->
+             inner rest (List.append errors [e]) in
```

Comment:
Could you use `e :: errors` here instead? `List.append` is relatively expensive since it duplicates the entire list (which isn't vey big here, but still). You can use `List.rev` before printing to get the same order.

Reply:
Great! Thanks :)